src/Microsoft.AspNetCore.Mvc.Core/Builder/MvcEndpointInfo.cs

```diff
@@ -5,106 +5,101 @@
-
-            if (constraints != null)
-            {
-                foreach (var kvp in constraints)
-                {
-                    constraintBuilder.AddConstraint(kvp.Key, kvp.Value);
-                }
-            }
+            var constraints = new Dictionary<string, IList<IRouteConstraint>>(StringComparer.OrdinalIgnoreCase);
 
-            foreach (var parameter in parsedTemplate.Parameters)
+            foreach (var parameter in ParsedPattern.Parameters)
             {
-                if (parameter.IsOptional)
+                foreach (var parameterPolicy in parameter.ParameterPolicies)
                 {
-                    constraintBuilder.SetOptional(parameter.Name);
-                }
+                    var createdPolicy = parameterPolicyFactory.Create(parameter, parameterPolicy);
+                    if (createdPolicy is IRouteConstraint routeConstraint)
+                    {
+                        if (!constraints.TryGetValue(parameter.Name, out var paramConstraints))
+                        {
+                            paramConstraints = new List<IRouteConstraint>();
+                            constraints.Add(parameter.Name, paramConstraints);
+                        }
 
-                foreach (var inlineConstraint in parameter.InlineConstraints)
-                {
-                    constraintBuilder.AddResolvedConstraint(parameter.Name, inlineConstraint.Constraint);
+                        paramConstraints.Add(routeConstraint);
+                    }
                 }
             }
 
-            return constraintBuilder.Build();
+            return constraints;
         }
 
+        public string Name { get; }
+        public string Pattern { get; }
+
+        // Non-inline defaults
+        public RouteValueDictionary Defaults { get; }
+
+        // Inline and non-inline defaults merged into one
+        public RouteValueDictionary MergedDefaults { get; }
+
+        public IDictionary<string, IList<IRouteConstraint>> Constraints { get; }
+        public RouteValueDictionary DataTokens { get; }
+        internal RoutePattern ParsedPattern { get; private set; }
+
         private static RouteValueDictionary GetDefaults(
-            RouteTemplate parsedTemplate,
+            RoutePattern parsedTemplate,
             RouteValueDictionary defaults)
         {
             var result = defaults == null ? new RouteValueDictionary() : new RouteValueDictionary(defaults);
 
             foreach (var parameter in parsedTemplate.Parameters)
             {
-                if (parameter.DefaultValue != null)
+                if (parameter.Default != null)
                 {
                     if (result.TryGetValue(parameter.Name, out var value))
                     {
-                        if (!object.Equals(value, parameter.DefaultValue))
+                        if (!object.Equals(value, parameter.Default))
```

Comment:
@rynowak So the reason defaults are merged here is `defaults` already contains any inline defaults because the value is coming from the `IRouteBuilder` which already merges them. To get around this I only error if there is a duplicate default for a parameter and they are different. What do you think about RoutePattern doing this instead of erroring on any duplicate? Then I can remove this code.

Reply:
I think it's fine to allow duplicates in RoutePattern if they are the same **value**